src/runtime/pybuffer.cs

```diff
@@ -0,0 +1,167 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Python.Runtime
+{
+    public class PyBuffer
+    {
+        private IntPtr _handle;
+        private Runtime.Py_buffer _view;
+        private GCHandle _gchandle;
+
+        internal PyBuffer(IntPtr obj, int flags, out bool success)
+        {
+            int size = Marshal.SizeOf(typeof(Runtime.Py_buffer));
+            byte[] rawData = new byte[size];
+            _gchandle = GCHandle.Alloc(rawData, GCHandleType.Pinned);
+            _handle = _gchandle.AddrOfPinnedObject();
+            _view = (Runtime.Py_buffer)Marshal.PtrToStructure(_handle, typeof(Runtime.Py_buffer));
+
+            success = Runtime.PyObject_GetBuffer(obj, _handle, flags) >= 0;
+
+            if (_view.shape != IntPtr.Zero)
+            {
+                Shape = new long[_view.ndim];
+                Marshal.Copy(_view.shape, Shape, 0, (int)_view.len * sizeof(long));
+            }
+
+            if (_view.strides != IntPtr.Zero) {
+                Strides = new long[_view.ndim];
+                Marshal.Copy(_view.strides, Strides, 0, (int)_view.len * sizeof(long));
+            }
+
+            if (_view.suboffsets != IntPtr.Zero) {
+                SubOffsets = new long[_view.ndim];
+                Marshal.Copy(_view.suboffsets, SubOffsets, 0, (int)_view.len * sizeof(long));
+            }
+        }
+
+        public PyObject Object => new PyObject(_view.obj);
```

Comment:
Creating `PyObject` does not increment reference counter of the object. You have to manually do it in this getter.

Reply:
But PyObject_GetBuffer() does. And PyBuffer_Release decrements the reference count... Just read the docs first. https://docs.python.org/3/c-api/buffer.html

Reply:
But in this can you actually get a new reference, so you have to increase the recount, or cache the instance of pyobject in constructor.

Reply:
@SnGmng you are right, that it usually takes a pre-increfed pointer. But each instance requires its own incref. Otherwise when the instance is disposed, all other instances of `PyObject` pointing to the same reference will simply crash on access.